### ISSUE TITLE
Refactor the use of OpenSSL into a provider interface for PKI Cert Management

### ISSUE THREAD

**Issue by ebekker** [ebekker/ACMESharp#16]
As discussed [here](https://github.com/ebekker/letsencrypt-win/issues/12#issuecomment-154563451) and [here](https://github.com/ebekker/letsencrypt-win/issues/12#issuecomment-154581872), it would be nice to decouple the ACME library and clients from the OpenSSL library.

We can convert all the OpenSSL integration points into a well-defined interface of dependencies on PKI certificate management calls, and then refactor the current use of OpenSSL assembly and native library into the first implementation of this new interface.

Thereafter, we can provide an alternative implementation that is based on the well-defined and commonly used [OpenSSL CLI](https://www.openssl.org/docs/manmaster/apps/openssl.html) interface (i.e. calling into a [console binary](https://www.openssl.org/community/binaries.html)).  This will allow us to resolve the issue with 32-bit/64-bit compatibility, and also allow end-users to substitute the console binary with any other CLI-compatible program (e.g. a different version of the OpenSSL console binary, or an alternative implementation, such as [LibreSSL](http://www.libressl.org/)).

Finally, by abstracting out the cert management interface, we can make use of completely different implementations such as one that is based on the Windows-native [CertEnroll API](https://msdn.microsoft.com/en-us/library/windows/desktop/aa374863(v=vs.85).aspx).

**Comment by ebekker** [comment]
@lennybacon, I'm working on this feature now, and hopefully will have something committed soon so you can work on the CertEnroll version.

**Comment by matbech** [comment]
@ebekker, @lennybacon: To safe you some time, I have actually implemented a working "CertEnroll" version here which already supports SAN:
https://github.com/matbech/letsencrypt-win-simple/commit/375109970df5d10f724d6f7c7ab5f48b58ba9d27
See the GeneratePrivateKey: https://github.com/matbech/letsencrypt-win-simple/commit/375109970df5d10f724d6f7c7ab5f48b58ba9d27#diff-6fc5b584da198c8d64d11e15eceed859R302
and https://github.com/matbech/letsencrypt-win-simple/commit/375109970df5d10f724d6f7c7ab5f48b58ba9d27#diff-6fc5b584da198c8d64d11e15eceed859R327

**Comment by ebekker** [comment]
That's awesome -- I'm getting closing to finishing up my refactored interface to the Cert Management provider, so I'll push that up shortly, this should help to make it simple to create a CE version of the provider quickly.

**Comment by ebekker** [comment]
OK guys, I've just committed to the client-posh branch the changes to refactor the use of OpenSSL managed lib into an abstracted Cert Management provider interface.  If you would like to review it and give me some any feedback, that would be great.  The main class to start from is `LetsEncrypt.ACME.PKI.CertificateProvider`.

In refactoring, I've had to make the structure a bit more complicated than the simple CsrHelper class that was there before, but that was necessary to provide the necessary hooks to allow other implementations.

So the initial, and currently default provider is the `LetsEncrypt.ACME.PKI.Providers.OpenSslLibProvider` which provides backward compatibility with what was there before, just repackaged into the new provider interface.

My next step is to implement a second provider that uses OpenSSL CLI interface.
